### ISSUE TITLE
Figure out configuration

### ISSUE THREAD

**Issue by mholt** [timelinize/timelinize#4]
The application itself has configuration (e.g. server address); settings related to its use (e.g. obfuscation mode), and there's also repository-specific settings (i.e. that travel with the repository).

We need a UI to configure these things ([Tabler has a template for it](https://preview.tabler.io/settings.html)).

The application settings are stored in a JSON file.

The repo settings are stored in the database, in the `repo` table. It's a generic key-value store. Right now it only stores the repo ID and schema version.

**Comment by mholt** [comment]
Config brainstorms:

- It seems to me there are 3 layers of configuration:
   1) Timeline settings (data source options, etc.)
   2) User settings (color theme, etc.)
   3) Application settings (listener port, etc.)
- I'm wrestling with where to store (and in what format to store) the various configurations. It's a little annoying that there's 3 separate piles of settings but I think that's most correct.
- Timeline config could likely exist in the `repo` table of the DB (I would add more columns to give more first-class presence to values like entity IDs, data source IDs, etc.) -- I thought about just doing a JSON blob altogether but am not leaning towards that.
- User settings... man... initially I was thinking these should live on the computer, but if you go use Timelinize on your other machine, then all your user settings are lost. It'd be nice if you could either log in to a central server (that was the original plan before open sourcing), or... maybe your own self-hosted Tlz instance?
- Application settings probably belong in a config file on the local disk. This should be pretty typical/easy enough.


Before open sourcing, I had the idea of everyone having a Timelinize account. (The vast majority would have been free accounts.) Mainly for preserving your settings wherever you may use the app, without having to manually set it all up again or copy all the settings.
It would have also been a way to add public-data timelines to your app, so you could enrich your data views.

AND I even started brainstorming data share features that the login would facilitate. Not to store any of your data of course, just to coordinate permissions, authentications, etc.

I guess I could still make some closed-source cherry-on-top features like that as sort of a SaaS.

Or, people could host their own Tlz instances maybe. If you switch computers just login to your "main" instance on your home server. That kind of thing.

---

Anyway, the short of it is:

- Timeline settings need to live in the timeline. Currently leaning towards rows in the DB but also considered a JSON file or blob.
- User settings ideally should follow the user around from device to device, or at least be accessible from multiple devices (either centrally or self-hosted).
- Application settings live on the computer in a JSON file and stay there.

:dizzy_face:

**Comment by rubiojr** [comment]
Thoughts:

* Could the app settings and the timeline settings be in a single, external file stored on the server? A json or a yaml file with a general section for app settings and then a sub section for each timeline. That would expose everything for admins to configure without having to provide tooling or ui initially, and reduce three different configuration layers to two
* It sounds reasonable to me for the user settings to live in an account table on the DB, so they are client device independent

**Comment by mholt** [comment]
@rubiojr Yes, sounds good! In fact I think that's basically what I ended up doing.

I made a lot of progress on this over the holidays.

- Application settings are stored in a local config.json file. It can now be changed through the UI and API endpoints!
- Timeline settings (like default data source options) have a place in the DB schema, for now, but they aren't implemented yet. I will likely have to revise the schema when I revisit this, but it's not needed for now. Data source options have reasonable defaults and can be changed per-import at import-time for the time being.
- I don't have much by way of "user settings" yet, I'm just wrapping them into "Application settings" for now. The reason being, I don't really have a concept of a "user" yet, like, there's no accounts or authentication. Maybe later.

So, in the long run, I still need to figure out "user settings" and, by necessity, user accounts, etc... but for now I think this issue is done.
